export type SettingType = "Slider" | "Binding" | "Checkbox"
type SettingCategory = "Controls" | "Gameplay" | "Graphics"
type ActionType = "Mouse" | "Controller" | "Keyboard"

export type SettingInfo = {
	Title: string,
	Description: string?,
	Category: SettingCategory,
	Folder: string?,
	Type: SettingType,
	Value: any,
	Default: any,
	Max: number?,
	Min: number?,
	Step: number?,
	ActionTypes: { ActionType }?,
}

local Info = {
	ShowFPS = {
		Title = "Show Framerate",
		Category = "Gameplay",
		Description = "Show game framerate in bottom left corner of screen",
		Type = "Checkbox",
		Default = false,
	},
	OneHandedSteering = {
		Title = "One Handed Steering",
		Category = "Gameplay",
		Description = "Only one hand will be on the steering wheel.",
		Type = "Checkbox",
		Default = false,
	},
	ShowInputs = {
		Title = "Show Inputs",
		Category = "Gameplay",
		Description = "Show vehicle inputs in bottom right corner of screen",
		Type = "Checkbox",
		Default = false,
	},
	ControllerDZone = {
		Title = "Controller Deadzone",
		Category = "Controls",
		Type = "Slider",
		Default = 0,
		Max = 0.9,
	},
	MSteerSensivity = {
		Folder = "Mouse Steering Settings",
		Title = "Sensivity",
		Category = "Controls",
		Description = "The speed of cursor in mouse steering mode",
		Type = "Slider",
		Default = 1,
		Min = 0.05,
		Max = 4,
	},
	MSteerDZone = {
		Folder = "Mouse Steering Settings",
		Title = "Steering Deadzone",
		Category = "Controls",
		Description = "Deadzone in mouse steering mode",
		Type = "Slider",
		Default = 0,
		Max = 0.9,
	},
	MSteerCursorY = {
		Folder = "Mouse Steering Settings",
		Title = "Cursor Y",
		Category = "Controls",
		Description = "The Y of cursor in mouse steering mode",
		Type = "Slider",
		Default = 0.2,
		Max = 1,
	},
	MSteerWidth = {
		Folder = "Mouse Steering Settings",
		Title = "Steering Width",
		Category = "Controls",
		Description = "Steering width in mouse steering mode",
		Type = "Slider",
		Default = 0.7,
		Min = 0.1,
		Max = 1,
	},
	ThrottlePressTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Throttle Press Time",
		Category = "Controls",
		Description = "Time to fully press the throttle in ms",
		Type = "Slider",
		Default = 200,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	ThrottleReleaseTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Throttle Release Time",
		Category = "Controls",
		Description = "Time to fully release the throttle in ms",
		Type = "Slider",
		Default = 100,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	BrakePressTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Brake Press Time",
		Category = "Controls",
		Description = "Time to fully press the brake in ms",
		Type = "Slider",
		Default = 300,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	BrakeReleaseTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Brake Release Time",
		Category = "Controls",
		Description = "Time to fully release the brake in ms",
		Type = "Slider",
		Default = 200,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	ClutchPressTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Clutch Press Time",
		Category = "Controls",
		Description = "Time to fully press the clutch in ms",
		Type = "Slider",
		Default = 100,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	ClutchReleaseTime = {
		Folder = "Keyboard Pedals Settings",
		Title = "Clutch Release Time",
		Category = "Controls",
		Description = "Time to fully release the clutch in ms",
		Type = "Slider",
		Default = 300,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
} :: { [string]: SettingInfo }

return Info
