export type SettingType = "Slider" | "Binding" | "Checkbox"
type SettingCategory = "Controls" | "Gameplay" | "Graphics"
type ActionType = "Mouse" | "Controller" | "Keyboard"

export type SettingInfo = {
	Name: string,
	Title: string,
	Description: string?,
	Category: SettingCategory?,
	Folder: string?,
	Type: SettingType,
	Value: any,
	Default: any,
	Max: number?,
	Min: number?,
	Step: number?,
	ActionTypes: { ActionType }?,
}

export type FolderInfo = {
	Name: string,
	Title: string,
	Description: string?,
	Category: SettingCategory?,
	Folder: string?,
}

local Folders = {
	{
		Name = "MSteer",
		Title = "Mouse Steering Settings",
		Category = "Controls",
	},
	{
		Name = "KPedals",
		Title = "Keyboard Pedals Settings",
		Category = "Controls",
	},
} :: { FolderInfo }

local Settings = {
	{
		Name = "ShowFPS",
		Title = "Show Framerate",
		Category = "Gameplay",
		Description = "Show game framerate in bottom left corner of screen",
		Type = "Checkbox",
		Default = false,
	},
	{
		Name = "OneHandedSteering",
		Title = "One Handed Steering",
		Category = "Gameplay",
		Description = "Only one hand will be on the steering wheel.",
		Type = "Checkbox",
		Default = false,
	},
	{
		Name = "ShowInputs",
		Title = "Show Inputs",
		Category = "Gameplay",
		Description = "Show vehicle inputs in bottom right corner of screen",
		Type = "Checkbox",
		Default = false,
	},
	{
		Name = "ControllerDZone",
		Title = "Controller Deadzone",
		Category = "Controls",
		Type = "Slider",
		Default = 0,
		Max = 0.9,
	},
	{
		Name = "MSteerSensivity",
		Folder = "MSteer",
		Title = "Sensivity",
		Description = "The speed of cursor in mouse steering mode",
		Type = "Slider",
		Default = 1,
		Min = 0.05,
		Max = 4,
	},
	{
		Name = "MSteerDZone",
		Folder = "MSteer",
		Title = "Steering Deadzone",
		Description = "Deadzone in mouse steering mode",
		Type = "Slider",
		Default = 0,
		Max = 0.9,
	},
	{
		Name = "MSteerWidth",
		Folder = "MSteer",
		Title = "Steering Width",
		Description = "Steering width in mouse steering mode",
		Type = "Slider",
		Default = 0.7,
		Min = 0.1,
		Max = 1,
	},
	{
		Name = "MSteerExp",
		Folder = "MSteer",
		Title = "Steering Exponential",
		Description = "Steering exponential degree, recommended values: 1.00-1.20",
		Type = "Slider",
		Default = 1.0,
		Min = 0.8,
		Max = 1.5,
	},
	{
		Name = "MSteerCursorY",
		Folder = "MSteer",
		Title = "Cursor Y",
		Description = "The Y of cursor in mouse steering mode",
		Type = "Slider",
		Default = 0.2,
		Max = 1,
	},
	{
		Name = "ThrottlePressTime",
		Folder = "KPedals",
		Title = "Throttle Press Time",
		Description = "Time to fully press the throttle in ms",
		Type = "Slider",
		Default = 200,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	{
		Name = "ThrottleReleaseTime",
		Folder = "KPedals",
		Title = "Throttle Release Time",
		Description = "Time to fully release the throttle in ms",
		Type = "Slider",
		Default = 100,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	{
		Name = "BrakePressTime",
		Folder = "KPedals",
		Title = "Brake Press Time",
		Description = "Time to fully press the brake in ms",
		Type = "Slider",
		Default = 300,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	{
		Name = "BrakeReleaseTime",
		Folder = "KPedals",
		Title = "Brake Release Time",
		Description = "Time to fully release the brake in ms",
		Type = "Slider",
		Default = 200,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	{
		Name = "ClutchPressTime",
		Folder = "KPedals",
		Title = "Clutch Press Time",
		Description = "Time to fully press the clutch in ms",
		Type = "Slider",
		Default = 100,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
	{
		Name = "ClutchReleaseTime",
		Folder = "KPedals",
		Title = "Clutch Release Time",
		Description = "Time to fully release the clutch in ms",
		Type = "Slider",
		Default = 300,
		Min = 50,
		Max = 1000,
		Step = 25,
	},
} :: { SettingInfo }

return {
	Settings = Settings,
	Folders = Folders,
}
